app/docs/md/patterns/form-validation.md

```diff
@@ -0,0 +1,281 @@
+
+---
+title: Form Validation
+---
+
+HTML forms are very powerful on their own.
+It does not take much markup to capture all kinds of data.
+But for most applications we need to validate the data sent to avoid user error or intentional misuse.
+This section covers validation best practices using Enhance. It includes:
+
+- Client/Server Validation
+- Handling Validation Problems with Session
+
+Proper form validation usually includes both server side and client side together.
+Client validation is fastest since it is closest to the user and can be done without posting data.
+Server validation is most reliable since it cannot be circumvented as easily.
+
+## Server Validation
+
+For a simple form we could add validation logic ad-hoc directly in the handler.
+But as the form grows that becomes unmaintainable.
+As a best practice we recommend adding a data schema and validating requests against that schema.
+
+## Data Validator
+
+The `@begin/validator` will validate a form response against a given JSON schema.
+The example below shows a how to integrate the validator into a data access layer that can be reused in multiple routes.
+
+First we install the validator into the project.
+```bash
+npm i @begin/validator
+```
+
+In the Data Access layer we add a data schema for books (`/app/models/schema/books.mjs`).
+
+
+```javascript
+// /app/models/schema/books.mjs
```

Comment:
Might useful to use `doc-code` for these code blocks to show the filename, i.e.:

```html
<doc-code filename="app/models/schema/books.mjs>
…
</doc-code>
```